consensus/system_contract/system_contract.go

```diff
@@ -0,0 +1,174 @@
+// signers set to the ones provided by the user.
+func New(ctx context.Context, config *params.SystemContractConfig, client sync_service.EthClient) *SystemContract {
+	ctx, cancel := context.WithCancel(ctx)
+	systemContract := &SystemContract{
+		config: config,
+		client: client,
+
+		ctx:    ctx,
+		cancel: cancel,
+	}
+	systemContract.Start()
+	return systemContract
+}
+
+// Hardcoded ABI snippet for getSigners():
+//
+//	function getSigners() external view returns (uint64[] memory, address[] memory);
+const getSignersABI = `[
+  {
+    "inputs": [],
+    "name": "getSigners",
+    "outputs": [
+      { "internalType": "address[]", "type": "address[]" }
+    ],
+    "stateMutability": "view",
+    "type": "function"
+  }
+]`
+
+// fetchAllSigners calls getAllTransitions() on the contract, decodes the results,
+// and returns them as a slice of [StartBlock, Signer].
+func (s *SystemContract) fetchAllSigners(blockNum *big.Int) ([]SignerAddressL1, error) {
+	// Parse the ABI so we can encode and decode calls
+	parsedABI, err := abi.JSON(strings.NewReader(getSignersABI))
+	if err != nil {
+		return nil, fmt.Errorf("parse ABI error: %w", err)
+	}
+
+	// Pack the function call data
+	packed, err := parsedABI.Pack("getSigners")
+	if err != nil {
+		return nil, fmt.Errorf("ABI pack error: %w", err)
+	}
+
+	// Prepare the call
+	msg := ethereum.CallMsg{
+		To:   &s.config.SystemContractAddress, // The address of your SystemSignerRegistry contract
+		Data: packed,
+	}
+
+	// Perform the call via the ethclient interface (need to cast)
+	output, err := s.client.CallContract(s.ctx, msg, blockNum)
+	if err != nil {
+		return nil, fmt.Errorf("CallContract error: %w", err)
+	}
+	// Decode the (uint64[] memory, address[] memory) from output
+	var ret struct {
+		Blocks  []*big.Int
+		Signers []common.Address
+	}
+	err = parsedABI.UnpackIntoInterface(&ret, "getSigners", output)
+	if err != nil {
+		return nil, fmt.Errorf("unpack error: %w", err)
+	}
+
+	if len(ret.Blocks) != len(ret.Signers) {
+		return nil, fmt.Errorf("mismatched array lengths blocks=%d signers=%d",
+			len(ret.Blocks), len(ret.Signers))
+	}
+
+	// Convert them into []SignerAddressL1
+	result := make([]SignerAddressL1, 0, len(ret.Blocks))
+	for i := range ret.Blocks {
+		result = append(result, SignerAddressL1{
+			StartBlock: ret.Blocks[i].Uint64(),
+			Signer:     ret.Signers[i],
+		})
+	}
+	return result, nil
+}
+
+// Authorize injects a private key into the consensus engine to mint new blocks
+// with.
+func (s *SystemContract) Authorize(signer common.Address, signFn SignerFn) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	s.signer = signer
+	s.signFn = signFn
+}
+
+func (s *SystemContract) Start() {
+	log.Info("starting SystemContract")
+	go func() {
+		s.lock.Lock()
+		signers, err := s.fetchAllSigners(nil) // nil -> latest state
+		if err != nil {
+			log.Error("failed to get signer address from L1 System Contract", "err", err)
+		}
+		s.signerAddressesL1 = signers
+		s.lock.Unlock()
```

Comment:
Maybe this should even be done synchronously on startup and fail the node if it fails? if the node can't read the config it won't be able to do anything no?

Reply:
It won't, but it will try again in 10 seconds?